compiler/src/iree/compiler/Codegen/Common/GPU/GPUApplyPaddingLevel.cpp

```diff
@@ -61,22 +59,94 @@ getTiledOps(Operation *funcOp, IREE::GPU::TilingLevel tilingLevel) {
   return targets;
 }
 
+/// For each reduction dimension in the linalg operation `linalgOp`, get a tuple
+/// containing:
+///
+/// 1) loop index of `linalgOp` that is reduced,
+/// 2) an operand that is indexed by the reduction dimension,
+/// 3) dimension of operand that's reduced.
+///
+/// Example:
+///
+/// ```mlir
+/// #map = affine_map<(d0, d1) -> (d1, d0)>
+/// #map1 = affine_map<(d0, d1) -> (d0, d1)>
+/// #map2 = affine_map<(d0, d1) -> ()>
+/// [...]
+/// %0 = linalg.generic {indexing_maps = [#map, #map1, #map2],
+///                      iterator_types = ["reduction", "reduction"]}
+///                      ins(%arg0, %arg1 : tensor<?x?xf16>, tensor<?x?xf16>)
+///                      outs(%arg2 : tensor<f16>)
+/// ```
+///
+/// The above operation has 2 reduction dimensions (d0 and d1).
+///
+/// d0 corresponds to operand dimension 1 of %arg0, and
+/// d1 corresponds to operand dimension 0 of %arg0.
+///
+/// So [{0, %arg0, 1}, {1, %arg0, 0}] is returned.
+static FailureOr<SmallVector<std::tuple<unsigned, Value, unsigned>>>
+findReductionDims(linalg::LinalgOp linalgOp) {
+
+  SmallVector<AffineMap> indexingMaps = linalgOp.getIndexingMapsArray();
+  SmallVector<utils::IteratorType> iteratorTypes =
+      linalgOp.getIteratorTypesArray();
+
+  auto nReductionDims =
+      llvm::count(iteratorTypes, utils::IteratorType::reduction);
+  SmallVector<std::tuple<unsigned, Value, unsigned>> triplets;
+  triplets.reserve(nReductionDims);
+
+  for (unsigned loopIdx = 0; loopIdx < iteratorTypes.size(); ++loopIdx) {
```

Comment:
Don't recalculate the end value
```suggestion
  for (unsigned loopIdx = 0, endIdx = iteratorTypes.size(); loopIdx != endIdx; ++loopIdx) {
```